frontend/ui/widget/textwidget.lua

```diff
@@ -46,14 +57,44 @@ function TextWidget:updateSize()
     end
     self._updated = true
 
+    -- Compute height:
+    -- Used to be:
+    --   self._height = math.ceil(self.face.size * 1.5)
+    --   self._baseline_h = self._height*0.7
+    -- But better compute baseline alignment from freetype font metrics
+    -- to get better vertical centering of text in box
+    -- (Freetype doc on this at https://www.freetype.org/freetype2/docs/tutorial/step2.html)
+    local face_height, face_ascender = self.face.ftface:getHeightAndAscender()
+    self._height = math.ceil(face_height) + 2*self.padding
+    self._baseline_h = Math.round(face_ascender) + self.padding
+    -- With our UI fonts, this usually gives 0.72 to 0.74, so text is aligned
+    -- a bit lower than before with the hardcoded 0.7
+
     -- In case we draw truncated text, keep original self.text
     -- so caller can fetch it again
     self._text_to_draw = self.text
 
+    if self.text and type(self.text) ~= "string" then
```

Comment:
When does this happen?

Reply:
Should not, but can if we provide a number or else.
We'll see it displayed instead of crashing, and it simplify my next check:
`if not self.text or #self.text == 0 then self._is_empty = true self._length = 0`
It avoid having `nil` fixed/handled in lower level code, like in:
https://github.com/koreader/koreader/blob/e166a77eba30e84cd72a9854b289bd9466ecc200/frontend/ui/rendertext.lua#L210-L214
https://github.com/koreader/koreader/blob/e166a77eba30e84cd72a9854b289bd9466ecc200/frontend/util.lua#L311-L313

Reply:
Would something this work in this context? (That'd go underneath the method.) It's a convenient way to ensure conformity without bothering users.
```lua
dbg:guard(TextWidget, "updateSize",
    function()
        assert(type(self.text) == "string",
            "Wrong text type (expected string)")
    end)
```

Reply:
I can, it works as you expect (crash on the emulator). Same in setText(text) then.
I think I've seen some output from `logger.warn("renderUtf8Text called without text")` in the past, so it may happen we get a nil. It's ok if we crash then when that happens?
(Can't cause any crash while tapping around - but I'm sure we'll get some.)

Reply:
Assuming it only crashes when it comes across a problem so that the problem can be addressed, that's indeed the idea. ;-)

But if it's too big an issue atm then it'd probably be best to leave that for a follow-up PR.

Reply:
I've already added and tested it, so it will go :)
If you confirm dbg.guard() does not kick in nor add overhead on our devices, it's fine with me - as long as I get there the safety of my tostring() (othewise, XText Lua check will throw an error).

Reply:
Yeah, we possibly have a few `nil` `renderUtf8Text` lying around, that does ring a bell ;).

Reply:
No overhead, when debug is disabled it's just nothing:

https://github.com/koreader/koreader/blob/b21b4c8776393f57f2c8fc15f06d7b90b952556f/frontend/dbg.lua#L65

And when enabled:

https://github.com/koreader/koreader/blob/b21b4c8776393f57f2c8fc15f06d7b90b952556f/frontend/dbg.lua#L37-L49